spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -396,48 +397,48 @@ public void close() throws IOException {
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
+        if (spec.schema().columns().stream()
+            .noneMatch(j -> j.type().isStructType() && j.type().asStructType().field(i) != null)
+        ) {
+          idColumns.add(i);
+        }
+      }
 
-      // schema needed for the projection and filtering
       StructType sparkType = SparkSchemaUtil.convert(finalSchema);
       Schema requiredSchema = SparkSchemaUtil.prune(tableSchema, sparkType, task.residual(), caseSensitive);
       boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
       boolean hasExtraFilterColumns = requiredSchema.columns().size() != finalSchema.columns().size();
 
-      Schema iterSchema;
       Iterator<InternalRow> iter;
 
       if (hasJoinedPartitionColumns) {
-        // schema used to read data files
-        Schema readSchema = TypeUtil.selectNot(requiredSchema, idColumns);
         Schema partitionSchema = TypeUtil.select(requiredSchema, idColumns);
         PartitionRowConverter convertToRow = new PartitionRowConverter(partitionSchema, spec);
-        JoinedRow joined = new JoinedRow();
+        GenericInternalRow partition = (GenericInternalRow) convertToRow.apply(file.partition());
 
-        InternalRow partition = convertToRow.apply(file.partition());
-        joined.withRight(partition);
+        Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+        Map<String, Integer> partitionSpecFieldIndexMap = Maps.newHashMap();
+        for (int i = 0; i < spec.fields().size(); i++) {
+          partitionSpecFieldIndexMap.put(spec.fields().get(i).name(), i);
+        }
 
-        // create joined rows and project from the joined schema to the final schema
-        iterSchema = TypeUtil.join(readSchema, partitionSchema);
-        iter = Iterators.transform(open(task, readSchema), joined::withLeft);
+        for (Types.NestedField field : partitionSchema.columns()) {
+          int partitionIndex = partitionSpecFieldIndexMap.get(field.name());
+          partitionValueMap.put(field.fieldId(), partition.genericGet(partitionIndex));
+        }
 
+        iter = open(task, finalSchema, partitionValueMap);
       } else if (hasExtraFilterColumns) {
-        // add projection to the final schema
-        iterSchema = requiredSchema;
-        iter = open(task, requiredSchema);
-
+        iter = open(task, requiredSchema, Collections.emptyMap());
       } else {
-        // return the base iterator
-        iterSchema = finalSchema;
-        iter = open(task, finalSchema);
+        iter = open(task, finalSchema, Collections.emptyMap());
       }
-
-      // TODO: remove the projection by reporting the iterator's schema back to Spark
-      return Iterators.transform(iter,
-          APPLY_PROJECTION.bind(projection(finalSchema, iterSchema))::invoke);
+      return iter;
```

Comment:
I don't think we should remove the projection in this PR. There is still the case where the `requiredSchema` and the `finalSchema` are not the same.

I don't think that code path is ever actually used because Spark requests all columns needed to run filters and Iceberg tells Spark to run all of the filters (because Spark can't handle residuals per task). But to remove this, we need to also remove the `hasExtraFilterColumns` code path, which would mean adding an unrelated change to this PR and as well as adding more tests. So let's separate those into different changes and leave the projection here.